tools/oldest_constraints.txt

```diff
@@ -1,76 +1,80 @@
-# This file contains the oldest versions of our dependencies we say we require
-# in our packages or versions we need to support to maintain compatibility with
-# the versions included in the various Linux distros where we are packaged.
+# This file contains the oldest versions of our dependencies we're trying to
+# support. Usually these version numbers are taken from the packages of our
+# dependencies available in popular LTS Linux distros. Keeping compatibility
+# with those versions makes it much easier for OS maintainers to update their
+# Certbot packages.
+#
+# When updating these dependencies, we should try to only update them to the
+# oldest version of the package that is found in a non-EOL'd version of
+# CentOS, Debian, or Ubuntu that has Certbot packages in their OS repositories
+# using a version of Python we support. If the distro is EOL'd or using a
+# version of Python we don't support, it can be ignored.
 
 # CentOS/RHEL 7 EPEL constraints
-cffi==1.6.0
+# Some of these constraints may be stricter than necessary because they
+# initially referred to the Python 2 packages in CentOS/RHEL 7 with EPEL.
+cffi==1.9.1
 chardet==2.2.1
-configobj==4.7.2
 ipaddress==1.0.16
 mock==1.0.1
```

Comment:
> I'm struggling to follow along here, sorry for dumb questions :).

Please ask away! If you have a question, I think there's a good chance I made a mistake, I'm not explaining things well, etc. 

> Tthe oldest version of py3 mock appears to be 2.0. 

(As a somewhat separate point that you may be interested in, the reason this works is https://github.com/certbot/certbot/issues/7886 which has some follow up work that needs to be done that's described in https://github.com/certbot/certbot/issues/7913.)

> If EPEL7 has already packaged mock for py3, shouldn't we use that as the oldest version?

If the older version of our dependencies work, I didn't take the time to try and upgrade them now. I think when we have a desire to upgrade a specific dependency, we can look into it, see what versions are packaged, and make a decision about what versions we want to support, but trying to do this proactively doesn't feel like a great use of time to me personally. There's a longer description of the process/logic I used here at https://github.com/certbot/certbot/issues/8580#issue-780742942.

One assumption I'm making in this PR which may help to lay out explicitly is I'm assuming the versions of the Python 3.6 packages available for CentOS 7 + EPEL 7 are no older than their Python 2 packages. I think that's a reasonable assumption to make because Python 3.6 is newer and there are no Python 3.6 packages in the base CentOS/RHEL 7 system, but it's possible this isn't true. Verifying this seems really annoying though and if I'm wrong, I think I think the damage is pretty minimal and the EPEL maintainers can ask us to add support for an older version. If people are worried about this, what I'd like to do as a next step is check with our EPEL maintainers and see if this is OK before spending time rechecking every dependency declared in this file.